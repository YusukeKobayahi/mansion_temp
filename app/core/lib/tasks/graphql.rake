# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
