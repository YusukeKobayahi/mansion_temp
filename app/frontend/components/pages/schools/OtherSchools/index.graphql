fragment MsLibraryPrimarySchoolOtherSchools on PrimarySchool {
  city {
    name
    primarySchools {
      schoolId
      name
      mansionCount
    }
  }
}
