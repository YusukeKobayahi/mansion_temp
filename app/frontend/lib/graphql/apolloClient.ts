// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
