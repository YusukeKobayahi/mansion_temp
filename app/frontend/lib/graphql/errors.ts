export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
