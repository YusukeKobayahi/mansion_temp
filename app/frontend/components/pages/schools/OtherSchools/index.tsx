import { MsLibraryPrimarySchoolOtherSchoolsFragment } from "~/graphql/generated";

import styles from "~/components/pages/cities/OtherCities/index.module.scss";
import BoxHeading from "~/components/commons/headings/BoxHeading";
import LinkBlock from "~/components/commons/LinkBlock";
import { sumReducer } from "~/lib/utils";

export interface OtherSchoolsProps {
  data: MsLibraryPrimarySchoolOtherSchoolsFragment;
  schoolName: string;
}

const OtherSchools: React.FC<OtherSchoolsProps> = ({
  data,
  schoolName,
}: OtherSchoolsProps) => {
  if (
    !data.city ||
    !data.city.primarySchools ||
    data.city.primarySchools.length === 0
  )
    return null;
  const schoolsCount =
    data.city.primarySchools.map((school) => school.mansionCount).length > 0
      ? data.city.primarySchools.map((school) => school.mansionCount)
      : [0];
  const schoolsTotalCount = schoolsCount.reduce(sumReducer);
  const items = data.city.primarySchools.map((school) => (
    <LinkBlock
      key={school.schoolId}
      name={school.name + "区"}
      id={school.schoolId}
      mansionCount={school.mansionCount}
      link="schools"
      eventParams={{
        action: "msl_area_click",
        category: "msl_ichiran",
        label: "msl_syuhen",
        value: 1,
      }}
    />
  ));
  return schoolsTotalCount !== 0 ? (
    <section className={styles.container}>
      <h2>
        <BoxHeading>
          {schoolName}区周辺の小学校区から中古マンションを探す
        </BoxHeading>
      </h2>
      <ul className={styles.list}>{items}</ul>
    </section>
  ) : null;
};

export default OtherSchools;
