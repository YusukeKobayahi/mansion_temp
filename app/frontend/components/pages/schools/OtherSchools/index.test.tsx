import { render } from "@testing-library/react";
import mockData from "~/components/pages/schools/fixtures/mockData.json";
import { PrimarySchoolsQuery } from "~/graphql/generated";
import OtherSchools from ".";

describe("components/pages/schools/OtherSchools", () => {
  it("正常に表示される", () => {
    const mock = mockData as PrimarySchoolsQuery;
    const dummyData = {
      data: mock.primarySchools[0],
      schoolName: "笄小学校",
    };
    const { asFragment } = render(<OtherSchools {...dummyData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
