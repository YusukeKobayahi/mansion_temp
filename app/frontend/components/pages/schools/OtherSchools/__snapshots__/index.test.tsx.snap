// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/schools/OtherSchools 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <h2>
      <div
        class="heading"
      >
        笄小学校区周辺の小学校区から中古マンションを探す
      </div>
    </h2>
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30021"
        >
          御成門小学校区
          <span
            class="count"
          >
            (8件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30022"
        >
          芝小学校区
          <span
            class="count"
          >
            (5件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30023"
        >
          笄小学校区
          <span
            class="count"
          >
            (11件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30024"
        >
          麻布小学校区
          <span
            class="count"
          >
            (10件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30025"
        >
          南山小学校区
          <span
            class="count"
          >
            (2件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30026"
        >
          東町小学校区
          <span
            class="count"
          >
            (8件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30027"
        >
          本村小学校区
          <span
            class="count"
          >
            (6件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30028"
        >
          赤坂小学校区
          <span
            class="count"
          >
            (18件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30029"
        >
          青山小学校区
          <span
            class="count"
          >
            (1件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30030"
        >
          青南小学校区
          <span
            class="count"
          >
            (3件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30031"
        >
          芝浦小学校区
          <span
            class="count"
          >
            (9件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30032"
        >
          白金小学校区
          <span
            class="count"
          >
            (4件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30035"
        >
          御田小学校区
          <span
            class="count"
          >
            (6件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30036"
        >
          赤羽小学校区
          <span
            class="count"
          >
            (4件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30037"
        >
          高輪台小学校区
          <span
            class="count"
          >
            (11件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/schools/30038"
        >
          港南小学校区
          <span
            class="count"
          >
            (8件)
          </span>
        </button>
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
